hail/src/main/scala/is/hail/utils/RestartableByteArrayInputStream.scala

```diff
@@ -0,0 +1,69 @@
+package is.hail.utils
+
+import java.io.{ IOException, InputStream }
+
+// not thread safe
+class RestartableByteArrayInputStream (
+  start: Int,
+  private[this] var end: Int,
+  private[this] var buf: Array[Byte]
+) extends InputStream {
+  def this(buf: Array[Byte]) {
+    this(0, buf.length, buf)
+  }
+  def this() {
+    this(0, 0, null)
+  }
+  if (buf != null) {
+    require(start >= 0)
+    require(start <= end)
+    require(end <= buf.length)
+  }
+  private[this] var off: Int = start
+
+  override def read(): Int = {
+    if (off == end) {
+      return -1
+    }
+    val b = buf(off) & 0xff
```

Comment:
Is this equivalent to `buf(off).toInt`? If so, that seems clearer.

Reply:
When the JVM converts from `Byte` to `Int`, it preserves the sign of the value in addition to preserving the low bits. This is good if you intend to preserve the semantic value in two's complement, but bad if you want 0s in the high bits. The `buf(off)` is automatically coerced to an integer because of the bitwise operation, if it's more clear, I could change to `buf(off).toInt & 0xff`, which makes the implicit up-cast explicit and clearly indicates the `0xff` is doing something else.

```scala
scala> (0xf0.toByte)
res16: Byte = -16

scala> (0xf0.toByte).toInt
res17: Int = -16

scala> (0xf0.toByte).toInt == 0xf0
res18: Boolean = false

scala> ((0xf0.toByte).toInt & 0xff) == 0xf0
res19: Boolean = true
```

Reply:
Ah, right, thanks. I think it's fine how it is.